src/dbnode/storage/shard.go

```diff
@@ -1804,6 +1811,22 @@ func (s *dbShard) FetchBlocksMetadataV2(
 	return result, nil, nil
 }
 
+func (s *dbShard) PrepareBootstrap() error {
+	s.flushState.RLock()
+	flushStateBootstrapped := s.flushState.bootstrapped
+	s.flushState.RUnlock()
+	if flushStateBootstrapped {
+		return nil // Already prepared.
+	}
+
+	// Iterate flushed time ranges to determine which blocks are retrievable.
+	// NB(r): This must be done before bootstrap since during bootstrapping
+	// series will load blocks into series with series.LoadBlock(...) which
+	// needs to ask the shard whether certain time windows have been flushed or
+	// not.
+	return s.bootstrapFlushStates()
```

Comment:
`bootstrapFlushStates` again checks `s.flushState.bootstrapped`, potentially racey if something happens between the checks?

Reply:
Yeah, I think this is ok though, it locks when it actually sets values.